### PR TITLE
VS-54: Support Linq Query syntax

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**
Support LINQ Query Syntax for the MongoDB C#/.NET Analyzer

### CODE REVIEW

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/Linq/LinqQuery.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+// limitations under the License.
+//
+
+using System.Linq;
+using MongoDB.Analyzer.Tests.Common.DataModel;
+using MongoDB.Driver.Linq;
+
+namespace MongoDB.Analyzer.Tests.Common.TestCases.Linq
+{
+    public sealed class LinqQuery: TestCasesBase
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/Linq/LinqQuery.cs:22]: Let's name it `LinqQuerySyntax`

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 2]
```diff
@@ -60,32 +60,56 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
...
-        foreach (var node in root.DescendantNodes(descendToChildrenPredicate).OfType<InvocationExpressionSyntax>())
+        //LINQ Method Syntax
+        foreach (var node in root.DescendantNodes(descendToChildrenPredicate).OfType<ExpressionSyntax>())
         {
-            var methodSymbol = semanticModel.GetSymbolInfo(node).Symbol as IMethodSymbol;
-
-            if (!methodSymbol.IsDefinedInMongoLinq() ||
-                !methodSymbol.ReceiverType.IsIMongoQueryable() ||
-                !methodSymbol.ReturnType.IsIMongoQueryable())
+            if(node is not InvocationExpressionSyntax && node is not QueryExpressionSyntax)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:66]: Can this and the following `if` blocks be merged?

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 3]
```diff
@@ -60,32 +60,56 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
...
             {
                 continue;
             }
+            if(node is QueryExpressionSyntax queryNode)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:70]: Missing space here and other places.

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 4]
```diff
@@ -60,32 +60,56 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
             return !processedSyntaxNodes.Contains(n.Parent);
         };
 
-        foreach (var node in root.DescendantNodes(descendToChildrenPredicate).OfType<InvocationExpressionSyntax>())
+        //LINQ Method Syntax
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:63]: Need this comment?

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 5]
```diff
@@ -60,32 +60,56 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
...
+            }
+            
             processedSyntaxNodes.Add(node);
 
             // Find the deepest node that supplies IMongoQueryable symbol type
-            var deepestMongoQueryableNode = GetNextNestedInvocation(node);
-
-            while (deepestMongoQueryableNode != null)
+            var deepestMongoQueryableNode = node;
+            if(node is QueryExpressionSyntax queryExpressionSyntax)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:95]: This check is already done before.

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 6]
```diff
@@ -144,6 +168,45 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
...
+                            result = false;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:209]: Empty line missing.

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 7]
```diff
@@ -144,6 +168,45 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
     private static ExpressionSyntax GetNextNestedInvocation(ExpressionSyntax expressionSyntax) =>
         ((expressionSyntax as InvocationExpressionSyntax)?.Expression as MemberAccessExpressionSyntax)?.Expression;
 
+    private static bool PreanalyzeLinqQuerySyntax(SyntaxNode linqExpressionNode, SemanticModel semanticModel, List<InvalidExpressionAnalysisNode> invalidLinqExpressionNodes)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:171]: No way to combine with `PreanalyzeLinqExpression` ? Seems that they are almost the same.

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs [hunk 8]
```diff
@@ -34,7 +34,7 @@ internal static class SyntaxNodeExtensions
     public static MethodDeclarationSyntax GetSingleMethod(this SyntaxNode syntaxNode, string name) =>
         syntaxNode.DescendantNodes().OfType<MethodDeclarationSyntax>().Single(m => m.Identifier.Text == name);
 
-    public static SyntaxNode GetTopMostInvocationOrBinaryExpressionSyntax(SimpleNameSyntax identifier, IdentifierNameSyntax[] lambdaIdentifiers)
+    public static SyntaxNode GetTopMostInvocationOrBinaryExpressionSyntax(SimpleNameSyntax identifier, IdentifierNameSyntax[] lambdaIdentifiers, IdentifierNameSyntax [] queryIdentifiers = null)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:37]: Do we really need to distinguish between lambdaIdentifiers and queryIdentifiers ?

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/Linq/LinqQuery.cs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                        where (item.Age == 22 || item.Age == 25)
+                        where (item.LastName == "Doe" || item.Name == "John")
+                        where item.Address.EndsWith("Drive")
+                        select new { item.Age, item.Address };
+        }
+
+        [MQL("aggregate([{ \"$match\" : { \"$or\" : [{ \"Age\" : 22 }, { \"Age\" : 25 }] } }, { \"$match\" : { \"$or\" : [{ \"LastName\" : \"Doe\" }, { \"Name\" : \"John\" }] } }, { \"$match\" : { \"Address\" : /^Drive/s } }, { \"$project\" : { \"Age\" : \"$Age\", \"Address\" : \"$Address\", \"_id\" : 0 } }])")]
+        public void Query_Syntax_9()
+        {
+            var query = from item in GetMongoQueryable()
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/Linq/LinqQuery.cs:99]: Try having large difference between test cases, reflected in the test method name, for example `Query_Syntax_8` and `Query_Syntax_9` are practically identical.

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/Linq/LinqQuery.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+                        where (item.Age == 22 || item.Age == 25)
+                        where (item.LastName == "Doe" || item.Name == "John")
+                        where item.Address.StartsWith("Drive")
+                        select new { item.Age, item.Address, item.Name };
+        }
+
+        [MQL("aggregate([{ \"$match\" : { \"$or\" : [{ \"Age\" : 22 }, { \"Age\" : 25 }] } }, { \"$match\" : { \"$or\" : [{ \"LastName\" : \"Doe\" }, { \"Name\" : \"John\" }] } }, { \"$match\" : { \"Address\" : /^Drive/s } }, { \"$project\" : { \"Age\" : \"$Age\", \"Address\" : \"$Address\", \"Name\" : \"$Name\", \"_id\" : 0 } }])")]
+        public void Query_Syntax_11()
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/Linq/LinqQuery.cs:117]: Try eliminating the repeating prefix Query, and think about more meaningful names for test methods if possible.
Also need test cases for edge cases like:

1. Methods invocation `where Foo(something) `
2. `from item in q where Foo(item)`
3. Don't always use `queryable` name, as it matches the MQLGenerator variable.
4. Few queries in single methods
5. Not supported queries like `item.GetHashCode() == 123` and other.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 11]
```diff
@@ -60,33 +60,51 @@ public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalyzerContext cont
             return !processedSyntaxNodes.Contains(n.Parent);
         };
 
-        foreach (var node in root.DescendantNodes(descendToChildrenPredicate).OfType<InvocationExpressionSyntax>())
+        foreach (var node in root.DescendantNodes(descendToChildrenPredicate).OfType<ExpressionSyntax>())
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs]: Done

File: src/MongoDB.Analyzer/Core/Utilities/SymbolExtensions.cs [hunk 12]
```diff
@@ -93,6 +93,14 @@ public static bool IsContainedInLambda(this ISymbol symbol, SyntaxNode parentNod
         return result;
     }
 
+    public static bool ContainsQueryParameter(this ISymbol symbol, SyntaxNode parentNode)
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SymbolExtensions.cs]: Done
